### PR TITLE
Qt: new class ElidedLabel

### PR DESCRIPTION
**Pull request by deanlee**

### CODE REVIEW

File: selfdrive/ui/qt/util.h [hunk 1]
```diff
@@ -13,6 +13,14 @@ inline QString getBrandVersion() {
   return getBrand() + " v" + QString::fromStdString(Params().get("Version")).left(14).trimmed();
 }
 
+inline QString truncate(const QString& str, size_t width, bool show_ellipsis = true) {
+  if (str.length() > width) {
+    return show_ellipsis ? str.left(width) + "..." : str.left(width);
```
Comment (pd0wm) [inline selfdrive/ui/qt/util.h:18]:
```suggestion
    return show_ellipsis ? str.left(width) + "…" : str.left(width);
```

**Comment by pd0wm** [comment]
How did you choose 18? Isn't there space for more?

**Comment by deanlee** [comment]
maybe we should do truncate in AbstractControl later

**Comment by deanlee** [comment]
> How did you choose 18? Isn't there space for more?
can adjust to 22

**Comment by adeebshihadeh** [comment]
Maybe we should make a nice helper function to fit some text in a max width within a range of font sizes?

**Comment by deanlee** [comment]
maybe a helper function like this.but it's hard to use.will increase a lot of code

```
inline QString elideString(QWidget *w, const QString &str, int width = 0, Qt::TextElideMode mode = Qt::ElideRight) {
  QFontMetrics fm(w->font());
  return fm.elidedText(str, mode, width <= 0 ? w->width() : width);
}
```

File: selfdrive/ui/qt/offroad/settings.cc [hunk 2]
```diff
@@ -264,8 +264,8 @@ void SoftwarePanel::updateLabels() {
   updateBtn->setText("CHECK");
   updateBtn->setEnabled(true);
   gitBranchLbl->setText(QString::fromStdString(params.get("GitBranch")));
-  gitCommitLbl->setText(QString::fromStdString(params.get("GitCommit")).left(10));
-  osVersionLbl->setText(QString::fromStdString(Hardware::get_os_version()).trimmed());
+  gitCommitLbl->setText(QString::fromStdString(params.get("GitCommit")));
+  osVersionLbl->setText(QString::fromStdString(Hardware::get_os_version()));
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/settings.cc:268]: short hash is better than cutoff long one
```suggestion
  gitCommitLbl->setText(QString::fromStdString(params.get("GitCommit")).left(10));
  osVersionLbl->setText(QString::fromStdString(Hardware::get_os_version()).trimmed());
```

File: selfdrive/ui/qt/widgets/controls.cc [hunk 3]
```diff
@@ -85,3 +87,14 @@ ButtonControl::ButtonControl(const QString &title, const QString &text, const QS
   QObject::connect(&btn, &QPushButton::released, this, &ButtonControl::released);
   hlayout->addWidget(&btn);
 }
+
+void ElidedLabel::paintEvent(QPaintEvent *event) {
+  QFrame::paintEvent(event);
+
+  QPainter painter(this);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/controls.cc:94]: can't we just make a custom QLabel and override setText?

Comment (deanlee) [reply selfdrive/ui/qt/widgets/controls.cc:94]: We may not know the width of qlabel when calling settext

Comment (adeebshihadeh) [reply selfdrive/ui/qt/widgets/controls.cc:94]: we can use the showEvent

**Comment by deanlee** [comment]
override resizeEvent and paintEvent, so that it can be displayed correctly in all situation

**Pull request merged**